### PR TITLE
fix(core): prevent reinsertion of an entity referenced by a formula when it is removed from a collection

### PR DESCRIPTION
**Pull request by Ruby184**
### The Problem

When using a `toMany` relation on an entity and also adding a `formula` that references the same entity in the collection, removing the entity from the collection causes the `formula` relation to keep referencing it, resulting in the deleted entity being reinserted on flush.

### Description of the Solution

I updated `referencingProperties` to allow multiple references on the same model, not just the first. However, the problem persisted because even though `unsetIdentity` performs `delete helper(referrer).__data[prop.name];`, it does nothing since `__data` is an empty object. I added an explicit override of the reference with `referrer[prop.name] = null;`.

I’m not sure what the correct way to reset this reference is (another test is failing because of that), so advice would be appreciated. Options include:  
- Setting it to `null`  
- Setting it to `undefined`  
- Deleting the property with `delete referrer[prop.name]`

### CODE REVIEW

**Review by B4nan — COMMENTED** [review]
thanks!

**Comment by B4nan** [comment]
> Deleting the property with delete referrer[prop.name]

I think this would remove the get/set definition which we don't want.

Let me take a closer look why this is happening, maybe there is a better solution than changing this line.

**Comment by Ruby184** [comment]
Looks like the get / set for `formula` which is not bidirectional is not defined and this causes that empty data object and  maintaining reference to deleted model. If you look here https://github.com/mikro-orm/mikro-orm/blob/2c1741904b07613eee29ebe9c3929d4a5008da34/packages/core/src/entity/EntityHelper.ts#L106-L119
`isReference` is false and then it is skipped by second if statement because `persist` is false.

File: packages/core/src/unit-of-work/UnitOfWork.ts [hunk 1]
```diff
@@ -447,6 +447,7 @@ export class UnitOfWork {
           rel.removeWithoutPropagation(entity);
         } else if (rel === entity) {
           delete helper(referrer).__data[prop.name];
+          referrer[prop.name] = null;
```
[the remark this answers is not in the document]
Comment (Ruby184) [reply packages/core/src/unit-of-work/UnitOfWork.ts:450]: It looks like a side effect of not defined get/set for formula property.

**Comment by Ruby184** [comment]
@B4nan I have fixed the original problem by defining get/set for formulas, added another test case for formula without ref (which is passing) and lastly tried it with `mapToPk`, which seems to be wrong even when loaded - it is not PK but the whole object.

**Comment by B4nan** [comment]
> causes that empty data object

Well, I'd say the data object should not contain formulas. Feels weird to have the value there for relation properties but not for scalars. Formulas are read only, so we don't need propagation on them, I think the correct fix is the opposite, not defining the get/set on them, then we can call `delete referrer[prop.name]`, but conditionally, only for formulas.

**Comment by Ruby184** [comment]
Well, I have made changes according to your instructions, and the first two test cases are passing. However, for completeness, I added a third test case using `mapToPk`, and it seems it always adds a join on the formula. As a result, the `icon` relation is not mapped to the primary key, and the entire `Image` object is assigned instead.

I tried looking into how `mapToPk` works, and it appears to be handled by `set`, which we do not have for formulas. Do you have any thoughts on how we should handle this case?

**Comment by B4nan** [comment]
Using `mapToPk` with a relational formula sounds really weird on its own, no need to use a relation property for that if all you want is to select the value which your formula will give you.

Let's not try to fix multiple things in one PR, if there is a bug with `mapToPk`, let's create a separate repro for that first, we have some tests for that so it shouldn't be completely broken.

**Comment by Ruby184** [comment]
> Using `mapToPk` with a relational formula sounds really weird on its own, no need to use a relation property for that if all you want is to select the value which your formula will give you.

Yes, but I can see a use case for that. Imagine you want to select the primary key of an entity while keeping it in sync when the entity is removed, so you don’t want the deleted entity to appear in the serialized form.

> Let's not try to fix multiple things in one PR, if there is a bug with `mapToPk`, let's create a separate repro for that first, we have some tests for that so it shouldn't be completely broken.

Yeah, I tried to dig into that and have a reproduction and fix. I will probably add more tests since `mapToPk` is not very well covered by tests, and then I will open another PR.

**Comment by B4nan** [comment]
Will you have time to finish this in the upcoming days? I wanted to ship a new patch release.

**Comment by Ruby184** [comment]
> Will you have time to finish this in the upcoming days? I wanted to ship a new patch release.

Yes, I will finish that today.

**Comment by B4nan** [comment]
Ok, thanks. I already shipped the patch release yesterday evening, so no pressure :] Some of the fixes in master were getting a bit too old.

**Comment by Ruby184** [comment]
> Ok, thanks. I already shipped the patch release yesterday evening, so no pressure :] Some of the fixes in master were getting a bit too old.

Sorry for the delay. I have opened another PR to fix the population of `mapToPk` relations.

**Comment by Ruby184** [comment]
@B4nan This one is also ready for your review. Thanks.

**Review by B4nan — APPROVED** [review]
thanks!

**Pull request merged**